### ISSUE TITLE
`+=` results in weird object tree

### ISSUE THREAD

**Issue by wez** [mockersf/hocon.rs#13]
This works and produces an array containing a single object with 4 keys:

```
keys [{
  key: "1",
  mods: "CTRL|ALT",
  action: "ActivateTab",
  arg: "0"
}]
```

but this yields a weird resultant object which subsequently fails to deserialize when using [this workaround](https://github.com/mockersf/hocon.rs/issues/11#issuecomment-557603351)

```
keys += [{
  key: "1",
  mods: "CTRL|ALT",
  action: "ActivateTab",
  arg: "0"
}]
```

yields:

```
  "keys": [
    [
      {
        "key": "1"
      }
    ],
    [
      {
        "mods": "CTRL|ALT"
      }
    ],
    [
      {
        "action": "ActivateTab"
      }
    ],
    [
      {
        "arg": "0"
      }
    ]
  ],
```

I'm new to hocon, but I would expect both of these inputs to have the same outputs.

**Comment by wez** [comment]
Test case:

```rust
diff --git a/tests/parse.rs b/tests/parse.rs
index 44f1428..bbd0c05 100644
--- a/tests/parse.rs
+++ b/tests/parse.rs
@@ -684,7 +684,7 @@ fn parse_concat_arrays_with_plus_equal_with_init() {
 fn parse_concat_arrays_with_plus_equal_with_object() {
     let s = r#"{
         a += { b : 1 }
-        a += { b : 2 }
+        a += { b : 2, c : 3 }
     }"#;
     let doc: Hocon = dbg!(dbg!(HoconLoader::new().load_str(dbg!(s)))
         .expect("during test")
@@ -693,6 +693,7 @@ fn parse_concat_arrays_with_plus_equal_with_object() {

     assert_eq!(doc["a"][0]["b"].as_i64().expect("during test"), 1);
     assert_eq!(doc["a"][1]["b"].as_i64().expect("during test"), 2);
+    assert_eq!(doc["a"][1]["c"].as_i64().expect("during test"), 3);
 }

 #[test]
```

a workaround appears to be:

```
tmp = [{
  key: "1",
  mods: "CTRL|ALT",
  action: "ActivateTab",
  arg: "0"
}]

keys += ${tmp}
tmp = null
```

**Comment by mockersf** [comment]
thanks for finding this issue

I got it to work on my tests, see [here](https://github.com/mockersf/hocon.rs/commit/673983c303ef3b2bf6e74dd626ddd139f19a165c#diff-071984ce4a5a6b95613d3a89df10df5d) for the updated test with more cases.

Could you check if it work for you before release?

I'll probably release beginning of next week

**Comment by wez** [comment]
Thanks for the quick turnaround!

One issue is that when building with http fetching disabled, the uuid library is missing the v4 methods.  This change makes the build succeed (see https://docs.rs/uuid/0.8.1/uuid/#dependencies):

```
 $ git diff
diff --git a/Cargo.toml b/Cargo.toml
index e43f34e..c6568b9 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -21,7 +21,7 @@ java-properties = "1.2"
 memchr = "2.2"
 reqwest = { version = "0.9", optional = true }
 failure = "0.1"
-uuid = "0.8"
+uuid = {version="0.8", features=["v4"]}

 [dev-dependencies]
 test-generator = "0.3"
```

Trying this with something that looks like this:

```
keys += [
   { a: 1 },
   { a: 2 },
]
keys += [
   { a: 3 },
   { a: 4 },
]
```

produces:

```
keys = [
   [
     { a: 1 },
     { a: 2 },
   ],
   [
     { a: 3 },
     { a: 4 },
  ]
]
```

which is one more level of array nesting than I expected; I expected this result:

```
keys = [
     { a: 1 },
     { a: 2 },
     { a: 3 },
     { a: 4 },
]
```

am I phrasing my config file incorrectly or is the behavior still not quite right?

**Comment by wez** [comment]
I think the behavior might be correct now; it seems consistent with the equivalent operation in python, but I'm not sure how to append one array to another; is there a way to do that without doing `keys += { a: 3 }` for each element?  I have a feeling that the answer is no because the word `append` appears only once in the informal spec!

**Comment by mockersf** [comment]
I just checked the official Scala implementation of hocon, and for this config it gives the same result.

Trying a few other possibilities, there is:
```
keys : [ {a: 1}, {a: 2} ]
keys : ${keys} [ {a: 3}, {a: 4} ]
```
which both seems to give what you need

**Comment by wez** [comment]
Thanks for that suggestion; it looks promising, but I think there's another section of the code with a similar issue to the bit you fixed, because this:

```
keys : [ {a: 1, b:1}, {a: 2, b:2} ]
keys : ${keys} [ {a: 3, b:3}, {a: 4, b:4} ]
```

yields:

```
  "keys": [
    {
      "a": 1,
      "b": 1
    },
    {
      "a": 2,
      "b": 2
    },
    {
      "a": 3
    },
    {
      "b": 3
    },
    {
      "a": 4
    },
    {
      "b": 4
    }
  ],
```

**Comment by mockersf** [comment]
this one was harder to debug, but easier to fix! [This test](https://github.com/mockersf/hocon.rs/commit/0e1ee74719b8dab226b2e7ea0f8c5aa465e8cc1c#diff-071984ce4a5a6b95613d3a89df10df5d) works now so should be ok

**Comment by wez** [comment]
Thanks!

I think I found another issue though; here's a modified test case.
The issue is that intermediate assignments appear to have no effect:

```
 $ git diff
diff --git a/tests/parse.rs b/tests/parse.rs
index fb1d221..3aa526a 100644
--- a/tests/parse.rs
+++ b/tests/parse.rs
@@ -653,12 +653,14 @@ fn parse_concat_arrays_of_object_with_self_substitution() {
     let s = r#"{
         a : [ { a : 1, b : 2 }, { a : 2, b : 4 } ]
         a : ${a} [ {a : 3, b : 6}, {a : 4, b : 8 } ]
+        a : ${a} [ {a : 5, b : 10}, {a : 6, b : 12 } ]
     }"#;
     let doc: Hocon = dbg!(HoconLoader::new().load_str(dbg!(s)))
         .expect("during test")
         .hocon()
         .expect("during test");

+/*
     assert_eq!(doc["a"][0]["a"].as_i64().expect("during test"), 1);
     assert_eq!(doc["a"][0]["b"].as_i64().expect("during test"), 2);
     assert_eq!(doc["a"][1]["a"].as_i64().expect("during test"), 2);
@@ -667,6 +669,18 @@ fn parse_concat_arrays_of_object_with_self_substitution() {
     assert_eq!(doc["a"][2]["b"].as_i64().expect("during test"), 6);
     assert_eq!(doc["a"][3]["a"].as_i64().expect("during test"), 4);
     assert_eq!(doc["a"][3]["b"].as_i64().expect("during test"), 8);
+*/
+    let expect = r#"{
+        a : [ { a : 1, b : 2 }, { a : 2, b : 4 },
+              {a : 3, b : 6}, {a : 4, b : 8 },
+              {a : 5, b : 10}, {a : 6, b : 12 } ]
+    }"#;
+    let expect: Hocon = dbg!(HoconLoader::new().load_str(dbg!(expect)))
+        .expect("during test")
+        .hocon()
+        .expect("during test");
+
+    assert_eq!(doc, expect);
 }

 #[test]
```

**Comment by mockersf** [comment]
Ok, got that case fixed too... next ? 😄

**Comment by wez** [comment]
Thanks!  This works for me :-)
